Add runtime update policy (#454)
radicle-dev/radicle-registry: DEVELOPING.md

~~~
where `<revision>` is the new Git revision SHA.

Runtime updates
---------------

There are special policies and processes around updates to the `runtime` package.

Updates to the runtime are tracked by the `VERSION` exported from the `runtime`
crate. The updates fall into two categories: Implementation updates and

> specification updates.

Reviewer: We have been calling the latter semantic changes. can we settle on one terminology?

Author: Fixed in https://github.com/radicle-dev/radicle-registry/pull/454/commits/3e4c111a9e1a8d6fc40ca34aa58cded75f3be1d4

### Implementation updates

Implementation updates only change the implementation of the runtime but do not
affect the semantics.

Commits with implementation updates must increment the `impl_version` field of

> `VERSION`. They may not recompile `./runtime/latest.wasm`.

Reviewer: > They may not recompile `./runtime/latest.wasm`.

I think your communication goal with this bit is to say that such changes are not required to be submitted/released?

Author: That is only secondary. The primary goal is to test the native version against the original wasm runtime.

### Specification changes

> Specification changes must increment the `spec_version` field.

Reviewer 2: Should it reset the `impl_version` to `0`?

Author: There’s no need for this. But it probably is confusing keeping the same version for the implementation. I’ll change it and require a reset to zero.

Author: https://github.com/radicle-dev/radicle-registry/pull/454/commits/a40649eabbe7548a4f707a47d9dcbe8bf345a431

In a commit with a specification change you must also update the lastest Wasm
runtime.

```

> ./scripts/build-runtime-wasm ./runtime/latest.wasm

Reviewer 2: Don't we want to keep the older runtimes? It could be useful for debugging when things start crashing in the real world with dozens of versions coexisting.

Reviewer: That is a good point, although I don't think we could submit an older version once we bump the runtime version at each release, or can we? In any case, git always tracks each version and we are not rewriting to master so that should be given for free.

Reviewer 2: That's correct :+1:

```

For specification changes to take effect on an existing chain they need to be
deployed to the chain. This process has not been established yet.

> Special care must be taken that updated runtimes can handle state written be
> previous versions. A policy and process for this has not been established yet.

Reviewer 2: ```suggestion
Special care must be taken that updated runtimes can handle state written by
the previous versions. A policy and process for updates, which break that
compatibility have not been established yet.
```

Reviewer: ```suggestion
Changes to the chain state must be backwards-compatibility, requiring particular attention.
A policy and process for such updates have not been established yet.
```
